loadtest/control/gencontroller/config.go

```diff
@@ -39,6 +39,23 @@ type Config struct {
 	PercentDirectChannels float64 `default:"0.6" validate:"range:[0,1]"`
 	// The percentage of group channels to be created.
 	PercentGroupChannels float64 `default:"0.1" validate:"range:[0,1]"`
+
+	// Indicates the distribution of chanel members within channels.
+	ChannelsDistribution []ChannelsDistribution
+}
+
+// ChannelsDistribution holds the member limit and what percentage of channels will have the limit.
+// Probability indicates how frequently to choose a range. Max and Min percentage indicates the
+// range bounds of the channel list.
+//
+// For example, a probability of 0.8 and min/max of 0.0-0.3 for a channel array
+// of length 10 would mean that indices [0,3] would be chosen at a 80% probability chance.
+// And then the Memberlimit indicates the max member count for that range.
+type ChannelsDistribution struct {
+	MemberLimit   int64
+	MinIndexRange float64 `validate:"range:[0,1]"`
+	MaxIndexRange float64 `validate:"range:[0,1]"`
+	Probability   float64 `validate:"range:[0,1]"`
```

Comment:
Right, I guess I was trying to understand if we actually needed the extra layer of complexity introduced by the probability given we have limits in place. We can pick channels however we want I suppose, the distribution of members is given by the percentages and the associated limit. Unless I am misunderstanding the requirements here.

Reply:
Right. So after running the PR for couple times, what I noticed is that rarely does a channel gets filled up to its member limit. The other actions get finished before that. So essentially, what we end up with is something like this:

```
 user_count |         channelid
------------+----------------------------
         15 | dfyeai3tjbrydfsu16jb8nnxbh
         15 | g1qjoo1gafrctga5uwxspfc6aw
         15 | zgbyfa6g47nxzgeeucake46o1y
         15 | s9qtwquas38a7enfghzooph5na
         14 | ysr67sumh3bwpq1qmkzr1x4npr
         14 | 18gzpbkfuindfmtdw5h3fe738e
         10 | dtonbi1khfdjurjo7ptzu6y99o
         10 | wbj36pdofpn9ufnrbr6d9iicow
          1 | 9dcod9rwyi888rz6gwbd1oi9ka
          1 | fszzyu4og7nbdf86gqgjgf7rgo
          1 | aym7hnnu43d9xkgsp5z19hrt8c
          1 | qqz6rbnpfinexydb76m5xgw1uy
          1 | ryenezni9jnkdniuh6x1tjceqy
          1 | 179n3cut9fb8iki8ar7nq5qeiy
          1 | bb53b8mcqfdj7poefuji7x8njw
          1 | cg9hkb9cmpytdf58onodr93mje
          1 | 3dq6rb5ik7fa7rq1oasmf8ggjc
          1 | nh3ts7run385zepwgjb69qho1r
          1 | 96kqudmaktnz9bia1yyuf31m3a
          1 | c3gkze5imt8sfchkhutnh5oymr
(20 rows)
```

Which is an approximation towards the actual distribution. But none of the channels are actually fully filled up. Now if we were to remove the channel range selection, then all channels would be filled up uniformly without any priority. So we would lose the member distribution.

Your proposal would be fine if we were to fill up _every single channel_ up to their memberLimit. But that slightly removes the probabilistic nature of things. I am not too opinionated either ways though.

Reply:
> Here, we need to pick channels in group A at least 10 times as much as channels in group B, so the Probability of A should be 10 times greater than the Probability of B? But there are much much less channels in A than in B, and we specify that using the {Min,Max}IndexRange, as in [0, 0.03] for A and [0.03, 0.33] for B. Is that right?

Yep somewhat like that. It's a bit difficult to translate the mental model of the data we have in hand, directly into code.

For your example above, the index range and the memberLimit is a direct 1:1 map. For probability, we have to work it out with some estimation.

```
  "ChannelsDistribution":[
      {
        "MemberLimit": 5000, // for >1000 case. We can put any high-enough number here.
        "MinIndexRange": 0.0,
        "MaxIndexRange": 0.003, // for 0.3% of channels
        "Probability": 0.7 
// This is where there isn't a 1:1 translation and we have to work out some estimates ourselves. 
// We have to decide how much weight we want to give to a given channel band depending on how fast
// we want to fill up that channel range. In a way this would be proportional to the membercount. So if you want high membercount, place high probability in that band to make it fill up faster.
      },
      {
         ...
      }
    ]
```

Does that help?

Reply:
It does help, thank you!

I'm thinking that if probabilities should be proportional to the member limit (and I guess that proportional, maybe to a lesser extent, to the length of the range), maybe the code can compute those by itself, instead of letting the users provide that.

I'm also perfectly ok going with what we have here, testing it out, and if we see we're computing the same probabilities every time, modify that later.

Reply:
There's also one potential core issue affecting this logic which is that the only users we handle are those who are running in the load-test which may not be very accurate as we usually end up running a very small subset of all users to optimize speed so I suspect it may be quite easy to run out of users in a proper attempt.

Reply:
> I'm thinking that if probabilities should be proportional to the member limit (and I guess that proportional, maybe to a lesser extent, to the length of the range), maybe the code can compute those by itself, instead of letting the users provide that.

That's actually a decent idea. @streamer45 - can you confirm that `control.SelectWeighted` doesn't need the sum of the weights to be 100? If that's doable, then we can directly proxy the memberCount limit as a probability.

> There's also one potential core issue affecting this logic which is that the only users we handle are those who are running in the load-test which may not be very accurate as we usually end up running a very small subset of all users to optimize speed so I suspect it may be quite easy to run out of users in a proper attempt.

Yep exactly. I raised this in the channel discussion as well

>  But the way our load test works is a bunch of users log in and start generating data. And each controller runs from a given user's perspective. So that becomes a bit difficult.

Overall, I think this will be a best effort thing. More like an improvement over the current joinChannel functionality. We can't guarantee all channels are filled with the required limit of channel members. But atleast after the test is done, the channel member distribution should be close to what we want it to be.

Reply:
> @streamer45 - can you confirm that control.SelectWeighted doesn't need the sum of the weights to be 100?

As far as I remember is completely generic, the proportion between weights and their total sum determines the probability.

> Yep exactly. I raised this in the channel discussion as well

Apologies, I suppose I didn't fully understand what you meant there.

> Overall, I think this will be a best effort thing.

As long as we get on the same order of magnitute of channelmembers entries we are good but if that's not the case then I think we may need to reconsider the approach because it's an important table to have rightly sized.

Reply:
> As long as we get on the same order of magnitute of channelmembers entries we are good but if that's not the case then I think we may need to reconsider the approach because it's an important table to have rightly sized.

@agarciamontoro - how many users did it take to reach 300pps?

Reply:
We used 500 in that test, but I guess we can play with the number of users and their rate

Reply:
Ok, let me think some more on this. It's also possible that the tool can create users as needed and fill it up.